### ISSUE TITLE
Wrong use of  __sync_add_and_fetch() builtins.

### ISSUE THREAD

**Issue by yablacky** [pilight/pilight#262]
in branch rewrite, file libs/pilight/core/eventpool.c there is now following code:

```C
void eventpool_fd_enable_flush(struct eventpool_fd_t *node) {
        if(running() == 0) {
                return;
        }

        int nr = __sync_add_and_fetch(&node->doflush, 0);
        if(nr == 0) {
                __sync_add_and_fetch(&node->doflush, 1);
        }
}
void eventpool_fd_disable_flush(struct eventpool_fd_t *node) {
        if(running() == 0) {
                return;
        }

        int nr = __sync_add_and_fetch(&node->doflush, 0);
        if(nr == 1) {
                __sync_add_and_fetch(&node->doflush, -1);
        }
}
```

If the intention of using __sync buitins is that ```node->doflush``` will never be a value other than ```0``` or ```1```, then it does not work. Because the test and the set operations are split in two separate __sync builtins. As a result:
If ```node->doflush``` is ```0```, then two threads executing ```eventpool_fd_enable_flush()``` may render ```node->doflush``` to value ```2```.
If ```node->doflush``` is ```1```, then two threads running in ```eventpool_fd_disable_flush()``` may render ```node->doflush``` to value ```-1```.
This happens if both threads are done with the 1st __sync (the "test") and are checking the same unmodified old value of ```node->doflush``` before any of them runs into the 2nd __sync that modifies the value (and then both will modify it).

Test and set should be done within *one* __sync builtin like this:

```__sync_val_compare_and_swap(&node->doflush, 0, 1); // for enable```
```__sync_val_compare_and_swap(&node->doflush, 1, 0); // for disable```

Since intention of enable is to set ```node->doflush = 1``` and intention of disable is to set it ```0``` I suggest to ```eventpool_fd_enable_flush(struct eventpool_fd_t *node)``` really and simply doing a ```node->doflush = 1``` and ```eventpool_fd_disable_flush(...)``` a ```node->doflush = 0```.

**Comment by CurlyMoo** [comment]
> Since intention of enable is to set node->doflush = 1 and intention of disable is to set it 0 I suggest to eventpool_fd_enable_flush(struct eventpool_fd_t *node) really and simply doing a node->doflush = 1 and eventpool_fd_disable_flush(...) a node->doflush = 0.

Yes, but this is not threadsafe. We must either use expensive locks or CAS. Your compare and swap should work as well. Does your CAS suggestion also work when two threads want to either disable or enable simultaniously?

**Comment by yablacky** [comment]
What do you mean with threadsafeness in this functions?
The intention of enable and disable function in this implementation is to set a flag to 1 or 0 that's it. They do not care the flag's before value nor care (that is: tell someone) if they really changed the flag.

May be I'm wrong, then could you please give me an example (construct a case) where threadsafeness is an issue?

**Comment by CurlyMoo** [comment]
You mean set a `set` instruction is always threadsafe?

**Comment by yablacky** [comment]
If the set instruction is the only thing that it's all about (not tests before or after), yes.

And this is what ```eventpool_fd_enable_flush()``` is intended to do, namely set doflush = 1. None of the __sync builtins prevent that a different thread runs ```eventpool_fd_disable_flush()``` and sets doflush = 0 again before ```eventpool_fd_enable_flush()``` returns.

**Comment by yablacky** [comment]
To me it seems that ```doflush``` being a simple flag is a design flaw anyway. If there are several threads that modify the flag its a permanent race condition what actually happens. Lets suppose ```doflush``` is a counter. For example the number of threads that want ```doflush``` to be "true". Then a thread can be sure that flush is enabled after it called ```eventpool_fd_enable_flush()``` - regardless what other threads do. The last thread that calls ```eventpool_fd_disable_flush()``` turns flushing of again. This is only an example.

**Comment by CurlyMoo** [comment]
I see, the idea of flushing is making sure a buffer is cleaned before a socket is closed. That's why the flush enabling only happens in line 827 and 838. Nowhere else :smile: But we can make it a counter by implementing it as a non-blocking semaphore. Then we don't need CAS.

**Comment by yablacky** [comment]
```__sync_add_and_fetch()``` is a threadsafe counter - you don't need semaphores for a counter.

Doesn't close imply a flush? Anyhow, my considerations concerning use of __sync builtings in eventpool.c are not specific to enable/disable flush functions. I used that as an example only.

**Comment by CurlyMoo** [comment]
No, close flushes what is in the kernel socket buffer, not what is in the pilight buffer. So if i sent 100kb, 10kb is still not sent over sockets a close will make sure the 10kb will be flushed, but the 90kb that never received the actual `send` cannot be flushed because the kernel doesn't know about it. This flush will make sure all to-be-written data will be written before the socket is closed.

In regard to CAS. Can you explain how i can best implement a `check if value is ... then increase or decrease`. I that as simple as implement the compare and swap?

e.g.
```
i = 1;
if(i == 1) {
  i--
}
```
Does this imply that the i is only set to zero if the value is `1`?
```
i = 1;
__sync_val_compare_and_swap(&i, 1, 0);
```
Do this will be skipped?
```
i = 0;
__sync_val_compare_and_swap(&i, 1, 0);
```

And when should i use this?
```
i = 1;
while(!__sync_val_compare_and_swap(&i, 1, 0));
```

**Comment by yablacky** [comment]
To make is threadsafe, the test of a memory-location and its modification must be an atomic operation. Atomic means that only one thread at a time does the test and modification before a different thread does it on the same location.

This "atomicness" is what the __sync builtins provide.

What code you need to write depends on what you want to achieve.

```c
// despite theadsafeness:
if (i==1) i--;
// is the same as
if (i==1) i = 0;
// and if i can only be 0 or 1 then its the same as
i = 0;
// because: setting 0 to 0 is ok. If it was 1 then it has to be 0.
// Since the previous value of i is not of interesst here
i = 0;
// is a "threadsafe" operation.
```

**Comment by yablacky** [comment]
In cases where the flags should better be a counter (e.g. number of threads that want the flag being true, indicating "enable") just do it like this:
```c
void init_foobar(*context) {
    context->counter = 0;
}
void enable_foobar(*context) {
    __sync_add_and_fetch(&context->counter, 1);
}
void disable_foobar(*context) {
    __sync_add_and_fetch(&context->counter, -1);
}
bool is_enabled_foobar(*context) {
    return __sync_add_and_fetch(&context->counter, 0) > 0;
}
```

**Comment by yablacky** [comment]
Just because it was a little challenge for me to deduce that the original flag modification ```if (i==1) i--;``` is better written as a pure set operation ("assignment") which is threadsafe by itself.
```c
// next three lines are obviously equivalent, hopefully:
__lock_access_to(&i); if (i == 1) i--;              __unlock_access_to(&i);
__lock_access_to(&i); if (i == 1) i = i - 1;        __unlock_access_to(&i);
__lock_access_to(&i); if (i == 1) i = i - i;        __unlock_access_to(&i);
// can we written more general:
__lock_access_to(&i); if (i != 0) i = i - i;        __unlock_access_to(&i);
// and for 0 it is also correct:
__lock_access_to(&i); if (i == 0) i = i - i;        __unlock_access_to(&i);
// so both cases can be combined:
__lock_access_to(&i); if (i != 0 ||
                          i == 0) i = i - i;        __unlock_access_to(&i);
// which is:
__lock_access_to(&i); if (1) i = 0;                 __unlock_access_to(&i);
// and for short:
__lock_access_to(&i); i = 0;                        __unlock_access_to(&i);
```
For ```if(i == 0) i++``` it goes like this (spared the __un/lock here):
```c
if (i == 0) i++;
if (i == 0) i = i + 1;
if (i == 0) i = 0 + 1;
// for 1 also correct:
if (i == 1) i = 0 + 1;
// more general:
if (i != 0) i = 0 + 1;
// combined:
if (i == 0 || i != 0) i = 0 + 1;
// finaly and short and complete:
__lock_access_to(&i); i = 1;                        __unlock_access_to(&i);
```

Now assume ```i``` is a single global variable used by two threads, one of which is running
```__lock_access_to(&i); i = 0;                        __unlock_access_to(&i);```
the other "simultaneously" is running
```__lock_access_to(&i); i = 1;                        __unlock_access_to(&i);```

Due to the access lock either ```i = 0;``` is executed first or ```i = 1;``` is executed first (and the other assignment afterwards).

Now assume no ```__un/lock_access``` around the assignments: One thead runs
```i = 0;``` and the other runs
```i = 1;```
Even without access lock, either ```i = 0;``` is executed first or ```i = 1;``` is executed first (and the other assignment afterwards). With or without access lock makes no difference.

A prerequisite is that ```i``` is a data type short enough that can be written in one single CPU instruction, so that assigment to ```i``` is atomic and not interrupted. This depends on CPU but typically holds for C scalar data types like char, int. Longer data structures require access lock when assigning or fetching them in a consistent way. If really in doubt assignments can be replaced by:
```c
__sync_or_and_fetch (&i, 1);    // i = 1; (almost - should there be other bits set, they remain set)
__sync_and_and_fetch (&i, 0);   // i = 0;
```
q.e.d. :wink:

**Comment by CurlyMoo** [comment]
It should be fixed now.

**Comment by CurlyMoo** [comment]
I do again need your help.

I'm running the rewrite code myself but after a while i get a segfault on line 219 in `core/ntp.c`:
```
	if(cursor < nrservers) {
		logprintf(LOG_DEBUG, "trying to sync with ntp-server %s", ntpservers[cursor]);

		eventpool_socket_add("ping", ntpservers[cursor], 123, AF_INET, SOCK_DGRAM, 0, EVENTPOOL_TYPE_SOCKET_CLIENT, callback, NULL, NULL);
	} else {
```
Somehow, the `cursor` can become equal of higher then the `nrservers` variable.

The cursor is increased on line 104 with a CAS (in a different thread):
```
	if(cursor+1 >= nrservers) {
		cursor = 0;
	} else {
		__sync_add_and_fetch(&cursor, 1);
	}
```

Any idea how i can increase the cursor without having to use locks and to also make sure that line 219 is never reached with a `cursor` that is higher then the `nrservers`?

**Comment by wo-rasp** [comment]
I am a bit struggling to understand the code, let .
1. static int curser
2. static unsigned int nrservers
A) I do assume that an overflow condition never occurs

3. if(cursor+1 >= nrservers) {
B) Why not: if(cursor < nrservers) { __sync_add_and_fetch(&cursor, 1); } else { cursor = 0; }

C) To increment the curser you use the CAS, but what is with the instructions starting with the comparison until you reset it to 0, or more accurate, you reset it to 0 but then cursor may get incremented by one or more to an unknown value - by an unknown number of waiting threads - to any value between 0 and the (# of waiting threads -1), depending on race conditions.

Example: 
a) 3 threads waiting before _sync_add
b) 4th tread waiting after cursor = 0;
c) the 3 waiting threads do continue
c1) 1st waiting thread reads value at address of cursor, adds 1 to 0 = 1
c2) 2nd waiting thread reads value at address of cursor, adds 1 to 1 = 2
c3) 3rd waiting thread reads value at address of cursor, adds 1 to 2 = 3
d) 4th thread continues, as cursor was already set to 0, the value remains at 3
e) Result: cursor = 3;

The issue is: Each thread has its own set of CPU registers and in those registers the value of cursor is no longer a global static variable, but is local to the thread and thus preserved.

**Comment by CurlyMoo** [comment]
I think the issue was indeed the `cursor+1`.

**Comment by yablacky** [comment]
Sorry for being offline a long time unfortunately I will be very busy for even more weeks.

Most probably the problem is, that *more* than just one thread executes this part of code at the same time:

```c
    if(cursor+1 >= nrservers) {
        cursor = 0;
    } else {
        __sync_add_and_fetch(&cursor, 1);
    }
```
On ```cursor``` the "test" (within if) and the "set" operation (within else path) is not atomic. Two (or more) threads may all do the test (in a way that after the test a different thread becomes active to test the same value of ```cursor```) and therefore all threads may found ``cursor`` being low enough to go to the else path to increment it.

The test and the increment must be atomic to prevent a thread from testing the already testested but not yet incremented value.

Due to the "if" and the different values to be written to ```cursor``` the operation is relatively complex. By eliminating the "if" it could be written like this:
```cursor = (cursor + 1) % nrservers``` an increment operation followed by a modulo operation.
As far as I see there is no single builtin nor CAS operation that calculates this atomicly.

I see two solutions:
(1) Using a semaphore to implement the above if - then - else as a critical section:
```c
__lock_semaphore_to_modify_cursor();
if (++cursor >= nrservers) cursor = 0;
__unlock_semaphore_to_modify_cursor();
```
(2) Letting ```cursor``` run out of bounds by just using ```__sync_add_and_fetch(&cursor, 1);``` all the time (without if - then - else) and add code to all places where ```cursor``` is read like this: ```cursor % nrservers```. The value of this expression is 0 .. nrservers-1 (provided nrservers is not negative).

Another idea doing it like is:
```c
int tested_cursor;
if((tested_cursor = cursor) < nrservers) {
        logprintf(LOG_DEBUG, "trying to sync with ntp-server %s", ntpservers[tested_cursor]);
...
```
This leads to the question: Why this "if"? Is  ```cursor >= nrservers``` is a regular situation?

**Comment by CurlyMoo** [comment]
I think i already fixed it :smile:
